File: javaScript/containerWithMostWater.js
// 11. Container With Most Water 
// Total Accepted: 91349
// Total Submissions: 256633

// Difficulty: Medium
// Given n non-negative integers a1, a2, ..., an, where each represents a point at coordinate (i, ai). n vertical lines are drawn such that the two endpoints of line i is at (i, ai) and (i, 0). Find two lines, which together with x-axis forms a container, such that the container contains the most water.

// Note: You may not slant the container.

// Tags Array Two Pointers

/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function(height) {
    //corner case
    if (height === null || height.length < 2) {
    	return 0
    }

    var left = 0;
    var right = height.length - 1;
    var maxArea = Math.min(height[left], height[right]) * (right - left);

    while (left < right){
    	area = Math.min(height[left], height[right]) * (right - left);
    	if (area > maxArea) {
    		maxArea = area;
    	}
    	if (height[left] < height[right]) {
    		var t1 = left;
    		for (t1 = left + 1; t1 < right && height[t1] < height[left]; ++t1) {}
    		left = t1;
    	} else {
    		var t2 = right;
    		for (t2 = right - 1; t2 > left && height[t2] < height[right]; ++t2) {}
    		right = t2;
    	}
    }

    return maxArea
};

//自己的思路都po在python那个code里面了，下面是讨论区的讲解，也很明了，所以贴在下面
/*
    设置两个指针i, j, 一头一尾, 相向而行. 假设i指向的挡板较低, j指向的挡板较高(height[i] < height[j]).
    下一步移动哪个指针?
    -- 若移动j, 无论height[j-1]是何种高度, 形成的面积都小于之前的面积.
    -- 若移动i, 若height[i+1] <= height[i], 面积一定缩小; 但若height[i+1] > height[i], 面积则有可能增大.
    综上, 应该移动指向较低挡板的那个指针.
*/

//而且，在代码里面找到了我没有写出来的加速办法，也就是，如果里面的柱子比当前柱子还矮，面积是一定更小的，所以不用计算它的面积

File: python/ContainerWithMostWater.py
# Container With Most Water

# Description
# Given n non-negative integers a1, a2, ..., an, where each represents a point at coordinate (i, ai). n vertical lines are drawn such that the two endpoints of line i is at (i, ai) and (i, 0). Find two lines, which together with x-axis forms a container, such that the container contains the most water.

# Notice
# You may not slant the container.

# Example
# Given [1,3,2], the max area of the container is 2.

# Tags 
# Two Pointers Array

# Related Problems 
# Medium Trapping Rain Water 35 %

class Solution:
    # @param heights: a list of integers
    # @return: an integer
    def maxArea(self, heights):
        # write your code here
        left = 0
        right = len(heights) - 1
        maxArea = 0

        while (left < right):
        	leftH = heights[left]
        	rightH = heights[right]
        	area = min(leftH, rightH) * (right - left)
        	if area > maxArea:
        		maxArea = area
        	if (leftH <= rightH):
        		left += 1
        	else:
        		right -= 1
        return maxArea

# 这一题刚一拿到，也是有点懵，但是后来一想，maxArea = maxX * maxY
# 我其实就可以从maxX -> 1 遍历，寻找到在特定X长度下，Y的最大值。
# 再对应到这一题，也就是，先开始maxX，是取数组最左和最右的两条线，然后容器的高度是由最小的那根决定的，所以是maxX * min(left, right)
# 接下来，要考虑缩短一格X， 然后Y 的变化了，缩短一格的时候要考虑是从哪一头开始缩，
# 因为高的变化了，不会使整个容量再增加，因为已经由最低的那根决定了，所以要缩短x要从低的那一头开始缩短，如果left++后，高度比当前left高，那整体容量还有可能增加， 如果比当前left低的话，就再left++看看，有没有可能找到比当前left高的
# 接下来，在再缩短X的过程中，发现，左右指针只有找到比他们自己更高的bar才有可能升级容量，不然容量就是一开始左右指针里面最低的那个 * 左右指针的差了

# 上面代码AC了，Yeah！终于AC了一回……
# 而且根据上面细致地分析，应该还有加速的办法，就是如果下一个left或者下一个right都小于当前的，就可以直接跳过，不用考虑了


